### ISSUE TITLE
Сборка проекта для серверного програмиста

### ISSUE THREAD

**Issue by AndreyGladkov** [bem/bem-forum-content-ru#413]
Хотелось бы узнать нет ли готового инструмента по сборке проекта на БЭМ
Что бы из папки desktop.bundles прекладывать html, css, js и картинки в папку app
Я ее просто заливаю на хостинг и сразу видно верстку,
Сейчас начал писать task для  gulp, но возможно у моей задачи есть решение и более верное.
Слышал про desp но нечего не нашел где взять его.

**Comment by tadatuta** [comment]
@AndreyGladkov структура для деплоя обычно достаточно специфичная от проекта к проекту, поэтому чего-то универсального (кажется) нет. Решение с использованием `gulp` вполне хорошее.
Если в процессе возникнут какие-то сложности — попробую помочь.

**Comment by AndreyGladkov** [comment]
И так я вернулся к этой задачи для меня пока не понятно как склеить один js и один css из всех что есть на странице удалив повторяющиеся блоки кода, борьщик умеет так склевать?

**Comment by Guria** [comment]
Нет, борщик не для этого.
Скорее всего вам нужен merged-бандл:
https://github.com/enb-bem/enb-bem-techs/blob/master/docs/build-merged-bundle.md
См. также https://github.com/bem/bem-forum-content-ru/issues/352#issuecomment-89412374

**Comment by AndreyGladkov** [comment]
Так я юзаю bem tools а не enb, и написал task для gulp чтобы он перекладывал весь html в папку app, насколько я понимаю мне нужен  merged - бандл для gulp что бы в процессе перекладывания js и css собирать их в один файл, а так же  какой либо task  что бы менять пути у изображений на  те что я укажу.
Есть какие либо инструменты для такой реализации или же нужно перейти на enb и писать task для него что бы он собирал дополнительно все как я хочу?

**Comment by Guria** [comment]
gulp врядли сможет грамотно вырезать повторяющиеся блоки кода. Думаю лучше переходить на enb и настроить merged бандл по примерам. А вот переписывать пути к картинкам как раз умеет борщик и в качестве бонуса делать freeze статики.
bem-tools, как я понял, скоро будет просто набором инструментов, а для сборки использовать тот же ENB.

**Comment by tadatuta** [comment]
@AndreyGladkov https://ru.bem.info/tools/bem/bem-tools/customization/#Сборка-merged-бандла-раньше-также-назывался-common

**Comment by AndreyGladkov** [comment]
Вообщем что то я не понял установил зависимости
```
  "dependencies": {
    "bem-environ": "^1.4.0",
    "bower": "^1.3.12",
    "bower-npm-install": "^0.5.9"
  },
  "devDependencies": {
    "bem": "0.9.0",
    "bem-environ": "1.4.0",
    "bem-tools-autoprefixer": "0.0.3",
    "bower": "1.4.1",
    "bower-npm-install": "0.5.9",
    "enb-bem": "0.1.0-beta5",
    "borschik": "^1.3.2",
    "borschik-tech-cleancss": "^1.0.3",
    "enb": "^0.13.15",
    "enb-autoprefixer": "^0.2.2",
    "enb-bem-techs": "^1.0.3",
    "enb-bemxjst": "^1.3.4",
    "enb-borschik": "^1.4.0",
    "enb-diverse-js": "^0.1.0",
    "enb-modules": "^0.2.0",
    "enb-stylus": "^1.2.0",
    "gulp": "^3.8.11",
    "stylus": "0.50.0"
  },
  "scripts": {
    "start": "enb server",
    "postinstall": "bower-npm-install",
    "deps": "bower-npm-install --non-interactive"
  },
  "engines": {
    "node": ">=0.10.0"
  }
}
```
Сделал npm i
и вы итоге сборщик падает на deps.js
[failed] [desktop.bundles/grid/grid.deps.js] deps

В чем может быть беда?

**Comment by Guria** [comment]
Покажи desktop.bundles/grid/grid.deps.js

**Comment by AndreyGladkov** [comment]
http://pastebin.com/C5Nkcn2e

**Comment by alexbaumgertner** [comment]
Мой вариант: 
1) [использовать](https://github.com/alexbaumgertner/sails-bem-project-stub/blob/master/views/.enb/make.js#L107) `merged` – бандл, причем во всех 
desktop.bundles/page-name использовать не `_page-name.css` и `_page-name.js` а сразу `_merged.css` и `_merged.js`
2) Копировать собранные файлы в `package.json`
https://github.com/alexbaumgertner/sails-bem-project-stub/blob/master/package.json#L53-L54

```js
"scripts": {
    "copy-views": "npm run copy-css && npm run copy-js",
    "copy-css": "cp views/desktop.bundles/merged/_merged.css views/static/css/",
    "copy-js": "cp views/desktop.bundles/merged/_merged.js views/static/js/"
  }
```

**Comment by AndreyGladkov** [comment]
Вообщем проблема следующая, если убрать из уровней переопределения те уровни в которых блоки то enb все верно собирает, если же их оставить то падает как я писал выше.
Через bemtools все собирается и работает, куда копать кто что думает?

**Comment by AndreyGladkov** [comment]
Проблема в том что я не заюзал deps-old.

**Comment by AndreyGladkov** [comment]
Столкнулся с проблемой 
```
var  enbBemTechs = require('enb-bem-techs');

nodeConfig.addTech([enbBemTechs.mergeBemdecl, {
			sources: bemdeclFiles,
			target: 'merged.bemdecl.js'
		}]);
```
Выдает  Concurrent techs for target: merged.bemdecl.js, techs: "merge-bemdecl" vs "bemjson-to-bemdecl" Брал код по ссылке выше и поправил его под себя.

**Comment by tadatuta** [comment]
@AndreyGladkov 
> Проблема в том что я не заюзал deps-old

Если сборка ломается на `deps` вместо `deps-old`, то, скорее всего, где-то в блочных deps.js-файлах встречаются циклические зависимости. Такие ситуации `deps-old` разруливает относительно наугад, так что рано или поздно есть риск нарваться на ситуацию, когда порядок сборки окажется неверный и повлиять на него будет сложно. Так что пока проект еще небольшой, рекомендую найти циклические зависимости и исправить (при использовании `deps` в консоль должны сыпаться указания на нужные блоки).

> Выдает Concurrent techs for target: merged.bemdecl.js, techs: "merge-bemdecl" vs "bemjson-to-bemdecl"

вот [коммит](https://github.com/bem/project-stub/commit/7a98ffd97fc3b6418c7e4c5957252424fb87f78d), который включает сборку merged-бандла для `project-stub`. Предлагаю повторить для своего проекта максимально похоже.

**Comment by AndreyGladkov** [comment]
Да взял c project-stub твой коммит все работает, спасибо, 
В консоль сыпятся названия  бандлов, при использованни  deps.

**Comment by tadatuta** [comment]
@AndreyGladkov нет, здесь речь не о циклической зависимости. Скорее всего-то где-то в *.deps.js есть опечатка. Для отлова можно попробовать добавить `console.log` в node_modules/enb-bem-techs/lib/deps/deps-resolver.js вокруг 203 строки и потом поискать вхождения по проекту.
